### ISSUE TITLE
Scipy minimize fail to iterate

### ISSUE THREAD

**Issue by carlitador** [scipy/scipy#12869]
Hi,
I am litteraly going crazy with Scipy.minimize.

I am simply trying to fit a certain list to a set of parameters representing a bezier curve.

My input is the "curveToFitArray". I have an initial guess "x0" that I cut in half to have bezier points coordinates in "Y" and "Z". I want to find the parameters that minimize the error between the curve generated by my bezier parametrization and the real curve this way:
```
#!/home/charles/miniconda3/envs/default/bin/python
from scipy.optimize import minimize
from scipy.interpolate import interp1d
import numpy as np
from scipy.special import comb
def bernstein_poly(i, n, t):
	return comb(n, i) * ( t**(n-i) ) * (1 - t)**i

def bezier_curve(points, nTimes=50):
	nPoints = len(points)
	xPoints = np.array([p[0] for p in points])
	yPoints = np.array([p[1] for p in points])

	t = 0.5 * (1 - np.cos(np.linspace(0,np.pi,nTimes))) 
	polynomial_array = np.array([bernstein_poly(i, nPoints-1, t) for i in range(0, nPoints)])

	xvals = np.dot(xPoints, polynomial_array)
	yvals = np.dot(yPoints, polynomial_array)

	return [list(c) for c in zip(xvals,yvals)]
def bspline(p,scale=[1.0,1.0],U='Y',V='X',nPoints=50):
	point_list = [[u*scale[0],v*scale[1]] for u,v in zip(p[U],p[V])]
	BSpline = bezier_curve(point_list,nPoints)
	#pts_BSpline = np.array(BSpline).T
	return sorted(BSpline, key=lambda x: x[0])
    
def Distance(curve):
    distance = np.cumsum(np.sqrt(np.sum(np.diff(curve, axis=0)**2, axis=1)))
    distance = np.insert(distance, 0, 0)/distance[-1]
    return distance
def divide(curve, n):
    distance = Distance(curve)
    alpha = np.linspace(0, 1, n)
    interpolation = interp1d(distance, curve, kind='linear', axis=0)
    interp_sort = sorted([[s, *c] for s, c in zip(alpha, interpolation(alpha))], key=lambda x: x[0])
    return [c[1:] for c in interp_sort]

def parseParameters(x):
    d = {}
    d["Y"] = list(x[0:4])
    d["Z"] = list(x[4:])
    return d

def objective(x,curveToFit):

    parameters = parseParameters(x)
    
    curveEval = divide(bspline(p = parameters,U='Y',V='Z',nPoints=len(curveToFit)),len(curveToFit))
    curveEval = np.array(curveEval)
    curveToFit = np.array(curveToFit)

    print("curveToFit",curveToFit)
    print("curveEval",curveEval)
    error = np.sqrt((np.subtract(curveEval, curveToFit)** 2))
    print("error",error)
    error = np.sum(error)
    
    return error


curveToFit = [[0.0, 0.0], [0.08462454350227121, -0.0006072162233802024], [0.16920310091996732, -0.0034025456809973417], [0.2533683125896774, -0.012003143737867554], [0.337444, -0.021644]]

x0 = [0.0, 0.050616600000000005, 0.253083, 0.337444, 0.0, 0.0, -0.021644, -0.021644]

bounds = [[0.0, 0.0], [-0.015, 0.35244400000000004], [-0.015, 0.35244400000000004], [0.337444, 0.337444],[0.0, 0.0], [-0.036643999999999996, 0.015], [-0.036643999999999996, 0.015], [-0.021644, -0.021644]]

result = minimize(
    fun = objective,
    x0 = x0,
    bounds = bounds,
    method='L-BFGS-B', # L-BFGS-B, SLSQP
    args=(curveToFit),
    options={'disp':True}
)

print(result)
```

I had it working nicely a few months ago. Then did problably some modifications on my code without checking the fitting was still working. Now I cant remember what I changed to make it stop working...

The algorithm just doesnt iterate and return a result very close to x0 everytime

Thanks a lot

**Comment by mdhaber** [comment]
Without looking too closely at the particular problem, here are some thoughts:

- This doesn't smell like a bug yet - an algorithm not iterating on a particular problem is not necessarily a bug, as x0 could near a stationary point - so stackoverflow would be a better place to start. We use GitHub for bugs and feature requests.
- Try different methods
- Try a different guess, or, if there is a reason to suspect this is a good guess, add noise to the guess
- Try a [global optimizer](http://scipy.github.io/devdocs/optimize.html#global-optimization), e.g. [`differential_evolution`](http://scipy.github.io/devdocs/generated/scipy.optimize.differential_evolution.html#scipy.optimize.differential_evolution). Shouldn't be too hard to adapt your code to that.

I'll leave this open for now. If it doesn't get too much attention and there is no more evidence that it's a bug, I'll close. Sound good?

**Comment by carlitador** [comment]
Hi mdhaber,

I didnt say it was necessary a bug. The thing is I have been used the algorithm (with L-BFGS) forever and it suddenly stopped to work and I can't explain why. I have spent days on trying to solve the problem. It's certainly related to something I have changed or that has changed in scipy.

I just can't afford in terms of computational time to use a global optimizer.
I am just posting here because I run out of solutions to make this work again and I really need help on this

thank you

**Comment by andyfaff** [comment]
I'm guessing that the reason it doesn't iterate is because there are four parameters whose lower bounds are equal to the upper bounds. Can you factor those out? The code that calculates the gradients strictly obeys the bounds, and it's difficult to calculate a finite difference if the lower and upper bounds are equal.

**Comment by carlitador** [comment]
Hi andrew,
This is not the reason because I am using this method in many other applications and this always works. I can't do with another option because I need to have a fixed number of parameters.

**Comment by andyfaff** [comment]
What is the output of your example? Are any Exceptions raised?

**Comment by andyfaff** [comment]
```
>>> from scipy.optimize import minimize, rosen
>>> # last parameter bounds are equal
>>> bounds = [(0, 10), (0, 10), (2, 2)]

>>> minimize(rosen, (2, 2, 2), method='L-BFGS-B', bounds=bounds)
/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/scipy/optimize/_numdiff.py:519: RuntimeWarning: invalid value encountered in true_divide
  J_transposed[i] = df / dx
      fun: 402.0
 hess_inv: <3x3 LbfgsInvHessProduct with dtype=float64>
      jac: array([1602.00001865, 1202.00001723,           nan])
  message: b'ABNORMAL_TERMINATION_IN_LNSRCH'
     nfev: 8
      nit: 0
     njev: 2
   status: 2
  success: False
        x: array([2., 2., 2.])
```

**Comment by mdhaber** [comment]
@carlitador do you happen to know the last version of SciPy in which l-bfgs-b was working for you?

**Comment by carlitador** [comment]
> ```
> >>> from scipy.optimize import minimize, rosen
> >>> # last parameter bounds are equal
> >>> bounds = [(0, 10), (0, 10), (2, 2)]
> 
> >>> minimize(rosen, (2, 2, 2), method='L-BFGS-B', bounds=bounds)
> /Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/scipy/optimize/_numdiff.py:519: RuntimeWarning: invalid value encountered in true_divide
>   J_transposed[i] = df / dx
>       fun: 402.0
>  hess_inv: <3x3 LbfgsInvHessProduct with dtype=float64>
>       jac: array([1602.00001865, 1202.00001723,           nan])
>   message: b'ABNORMAL_TERMINATION_IN_LNSRCH'
>      nfev: 8
>       nit: 0
>      njev: 2
>    status: 2
>   success: False
>         x: array([2., 2., 2.])
> ```

Please understand this is not the issue. I am using this equal bounds for many other projects and its been working well forever.

Here is the output of my problem

```
/home/charles/miniconda3/envs/default/lib/python3.8/site-packages/scipy/optimize/_numdiff.py:519: RuntimeWarning: invalid value encountered in true_divide
  J_transposed[i] = df / dx
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =            8     M =           10

At X0         4 variables are exactly at the bounds

At iterate    0    f=  1.97255D-02    |proj g|=          NaN

At iterate    1    f=  1.97255D-02    |proj g|=          NaN

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
    8      1      3      3     0     6         NaN   1.973D-02
  F =   1.9725495590310269E-002

CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH             
      fun: 0.01972549559031027
 hess_inv: <8x8 LbfgsInvHessProduct with dtype=float64>
      jac: array([        nan, -0.0637971 , -0.06698817,         nan,         nan,
       -0.87670644, -0.93680716,         nan])
  message: b'CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH'
     nfev: 27
      nit: 1
     njev: 3
   status: 0
  success: True
        x: array([ 0.       ,  0.0506166,  0.253083 ,  0.337444 ,  0.       ,
        0.       , -0.021644 , -0.021644 ])
```

**Comment by mayou36** [comment]
So while this _may_ works, putting the values at the bounds is not what you want and it may gives some NaNs. What about you don't give the values to the function (but e.g. use `partial` to give them as constants)? While it does not "solve" the break in the code, I think setting the bounds like this is also happen to break for different reasons.

**Comment by vaniagolushko** [comment]
I experienced a very similar problem, at least I think so. So you probably know the book 
IPython Interactive Computing and Visualization Cookbook, Second Edition (2018), by Cyrille Rossant
Recently when testing the examples given there and in particular
[https://ipython-books.github.io/94-finding-the-equilibrium-state-of-a-physical-system-by-minimizing-its-potential-energy/](url)
I found that example doesn't work.
L-BFGS fails to fined minima of the problem with bounds, and just says that inital configuration is a solution to the problem.
When changing the method to Powell's everything works.
I message Cyrille and he told me that previously it was working, but now indeed only Powell's works (the program itself remains unchanged)
[https://github.com/ipython-books/cookbook-2nd-code/pull/4](url)
So it is quite possible that behavior of the method implementation has changed itself.

**Comment by marcoaqil** [comment]
i am getting this same issue with L-BFGS-B not progressing (NaN gradient) if any bounds have equal upper/lower, however I am sure the same approach was working in previous scipy versions with L-BFGS-B. [The docs generally suggests using bounds exactly this way to keep parameters fixed](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.Bounds.html#scipy.optimize.Bounds), so it would be great if it could be fixed again. thanks :)

**Comment by andyfaff** [comment]
I believe this issue should be fixed with #13096, with the fix being available in scipy 1.8.

If I understand correctly the issue arises because of problems estimating the gradient when a lower bound was equal to an upper bound. Changes made to strictly obey bounds during finite difference differentiation caused approx_derivative to start returning grads containing NaN. L-BFGS-B (and a couple of others) didn't like that. The fix in #13096 factorises parameters with equal bounds out.
